Introduce custom exceptions for Vulkan-Hpp (#97)

RippeR37 opened the pull request:
Base exception architecture:

* vk::Error
** vk::LogicError : std::logic_error
** vk::SystemError : std::system_error
*** vk::$ErrorResultEnumName$Error : SystemError

This is to have backward compatibility with existing code which was catching std::logic_error or std::system_error.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Hpp?pullRequest=99) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/Vulkan-Hpp?pullRequest=99) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/Vulkan-Hpp?pullRequest=99) it.</sub>

asuessenbach commented on VulkanHppGenerator.cpp:
```diff
@@ -62,6 +62,53 @@ const std::string exceptionHeader(
+  "  class Error\n"
+  "  {\n"
+  "    public:\n"
+  "    virtual const char* what() const noexcept = 0;\n"
```
I think, there's no need to have this function here, as both std::logic_error and std::system_error already provide that.
In order to make vk::Error a polymorphic type, it should be enough to have, for example, a virtual destructor here.

RippeR37 replied:
I added this, because - since we can't inherit from both `std::exception` and `std::system_error` - when user would want to catch all Vulkan exceptions at once, he couldn't do anything with it. Imagine:
```
try {
    // ...
} catch (/* something specific */) {
    // ...
} catch (const vk::Error& vk_error) {
    // What to do with `vk_error`? Can't even print what was the problem...
    // not much more useful then `catch (...)`
}
```
So I think that it would be nice to have at least `what()` available there. If you disagree with this argument I will remove this.

As for virtual destructor - nice catch, I should have seen this... :+1:

RippeR37 commented:
@asuessenbach 
Not to pollute history I've amended last commit and force-pushed changes (but I made sure to re-edit so new metadata would appear) - I hope it's fine.

Changes I made:
* Added virtual destructors for `vk::Error`, `vk::SystemError` and `vk::LogicError`,
* Removed trailing space from `throwResultException()` definition,
* Added `vulkan.hpp` generated from current code.

P.S. I seriously have no idea why you chose to use CRLF for `vulkan.hpp` and LF for `VulkanHppGenerator.cpp` - this is really confusing.

asuessenbach commented:
@RippeR37 would you mind to resolve the conflicts or your pull-request that occurred with the latest merge on the main branch?

RippeR37 commented:
@asuessenbach Since this cleanup was so big, I wanted to make sure that everything is okey so I've synced-up with current master and rebased my changes by hand on it. Basically everything is same, one function was deleted by this cleanup so my changes don't belong there anymore and I've added forward declaration of used functions to match what already is in code.

You can check it out, for me it looks okey.
